### PR TITLE
Updated SharpFont to 3.0.0, adjusted font rendering code for new fixed-point types

### PR DESCRIPTION
**Pull request by Robmaister**
Tested, main menu seems to be rendering fonts just fine.

If you'd prefer the fetch-thirdparty-deps scripts to be organized slightly differently, let me know. I just went with the minimal number of lines I'd have to change.

I'm also using the MSVC10 x86 version of freetype6.dll, if you wanted a different one also let me know.

### CODE REVIEW

**Comment by penev92** [comment]
We recently had problems with those on some machines, so this should be carefully tested.

File: thirdparty/fetch-thirdparty-deps-windows.sh [hunk 1]
```diff
@@ -16,10 +16,9 @@ fi
...
 if [ ! -f windows/freetype6.dll ]; then
 	echo "Fetching FreeType2 from nuget"
-	nuget install freetype2.redist -Version 2.4.11.3
-	cp ./freetype2.redist.2.4.11.3/bin/win32/zlib1.dll ./windows/
-	cp ./freetype2.redist.2.4.11.3/bin/win32/freetype6.dll ./windows/
-	rm -rf freetype2.redist.2.4.11.3
+	nuget install SharpFont.Dependencies -Version 2.5.5.1
+	cp ./SharpFont.Dependencies.2.5.5.1/bin/msvc10/x86/freetype6.dll ./windows/
+	rm -rf SharpFont.Dependencies.2.5.5.1
 fi
```
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: You will also need to replace/remove the `windows/zlib1.dll` below.

Comment (Robmaister) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: Wasn't sure if you needed zlib for anything else... I'll go ahead and remove it.

**Comment by Robmaister** [comment]
Zlib dependency removed, as requested.

**Comment by pchote** [comment]
This crashes under OSX with mono 3.12 and XQuartz 2.7.8_beta3:

```
Exception of type `System.OutOfMemoryException`: Out of memory
  at (wrapper managed-to-native) object:__icall_wrapper_mono_gc_alloc_vector (intptr,intptr,intptr)
  at (wrapper alloc) object:AllocVector (intptr,intptr)
  at OpenRA.Graphics.Sheet.CreateBuffer () [0x00048] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Sheet.cs:149 
  at OpenRA.Graphics.Sheet.GetData () [0x00002] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Sheet.cs:30 
  at OpenRA.Graphics.SpriteFont.CreateGlyph (Pair`2 c) [0x000e2] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/SpriteFont.cs:120 
  at OpenRA.Primitives.Cache`2[OpenRA.Primitives.Pair`2[System.Char,System.Drawing.Color],OpenRA.Graphics.GlyphInfo].get_Item (Pair`2 key) [0x00022] in /Users/paul/src/OpenRA/OpenRA.Game//Primitives/Cache.cs:39 
  at OpenRA.Graphics.SpriteFont.PrecacheColor (Color c, System.String name) [0x0004d] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/SpriteFont.cs:54 
  at OpenRA.Graphics.SpriteFont..ctor (System.String name, Int32 size, OpenRA.Graphics.SheetBuilder builder) [0x000a3] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/SpriteFont.cs:46 
  at OpenRA.Graphics.Renderer.<InitializeFonts>m__1 (KeyValuePair`2 x) [0x00029] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Renderer.cs:104 
  at System.Linq.Enumerable.ToDictionary[KeyValuePair`2,String,SpriteFont] (IEnumerable`1 source, System.Func`2 keySelector, System.Func`2 elementSelector, IEqualityComparer`1 comparer) [0x0002f] in /private/tmp/source-mono-mac-3.12.99-branch-32/bockbuild-mono-3.12.0-branch/profiles/mono-mac-xamarin/build-root/mono-3.12.1/mcs/class/System.Core/System.Linq/Enumerable.cs:2928 
  at System.Linq.Enumerable.ToDictionary[KeyValuePair`2,String,SpriteFont] (IEnumerable`1 source, System.Func`2 keySelector, System.Func`2 elementSelector) [0x00000] in /private/tmp/source-mono-mac-3.12.99-branch-32/bockbuild-mono-3.12.0-branch/profiles/mono-mac-xamarin/build-root/mono-3.12.1/mcs/class/System.Core/System.Linq/Enumerable.cs:2915 
  at OpenRA.Graphics.Renderer.InitializeFonts (OpenRA.Manifest m) [0x00034] in /Users/paul/src/OpenRA/OpenRA.Game//Graphics/Renderer.cs:103 
  at OpenRA.Game.InitializeMod (System.String mod, OpenRA.Arguments args) [0x00184] in /Users/paul/src/OpenRA/OpenRA.Game//Game.cs:307 
  at OpenRA.Game.Initialize (OpenRA.Arguments args) [0x002d0] in /Users/paul/src/OpenRA/OpenRA.Game//Game.cs:264 
  at OpenRA.Program.Run (System.String[] args) [0x00007] in /Users/paul/src/OpenRA/OpenRA.Game//Support/Program.cs:109 
  at OpenRA.Program.Main (System.String[] args) [0x00050] in /Users/paul/src/OpenRA/OpenRA.Game//Support/Program.cs:39 
```

**Comment by Robmaister** [comment]
hmmm... don't have a mac to test on. Going to poke around tomorrow afternoon and see if there's anything that could be eating up RAM internally.

**Comment by Mailaender** [comment]
Works on openSUSE Linux with Mono 3.12.1. :white_check_mark:

**Comment by x-a-n-a-x** [comment]
It works on Windows 8.1 with Visual Studio 2015 Preview.

File: thirdparty/fetch-thirdparty-deps.sh [hunk 2]
```diff
@@ -39,10 +39,10 @@ fi
...
 	echo "Fetching SharpFont from nuget"
-	nuget install SharpFont -Version 2.5.0.1
-	cp ./SharpFont.2.5.0.1/lib/net20/SharpFont* .
-	cp ./SharpFont.2.5.0.1/Content/SharpFont.dll.config .
-	rm -rf SharpFont.2.5.0.1
+	nuget install SharpFont -Version 3.0.0
+	cp ./SharpFont.3.0.0/lib/net20/SharpFont* .
+	cp ./SharpFont.3.0.0/config/SharpFont.dll.config .
+	rm -rf SharpFont.3.0.0
 fi
```
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps.sh]: You will also need to delete the `SharpFont.Dependencies.2.5.5.1` folder again.

**Comment by Robmaister** [comment]
Released SharpFont v3.0.1 which was just changing the X11 dependency on OS X and some minor documentation changes. Going to update the pull request and remove the conflicts in a little bit.

https://github.com/Robmaister/SharpFont/compare/v3.0.0...v3.0.1

**Comment by Robmaister** [comment]
Should be good to pull in now!

File: thirdparty/fetch-thirdparty-deps-windows.sh (hunk 1 above)
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: Go ahead then. =) It seems to be still there.

**Comment by pchote** [comment]
This now works under OSX, but the dlls cached in thirdparty need to be manually deleted first.

File: thirdparty/fetch-thirdparty-deps.sh [hunk 3]
```diff
@@ -39,10 +39,13 @@ fi
 
 if [ ! -f SharpFont.dll ]; then
 	echo "Fetching SharpFont from nuget"
-	nuget install SharpFont -Version 2.5.0.1
-	cp ./SharpFont.2.5.0.1/lib/net20/SharpFont* .
-	sed '/osx/s@\(dll="\)[^"]*\(" />\)@\1/Library/Frameworks/Mono.framework/Libraries/libfreetype.6.dylib\2@' ./SharpFont.2.5.0.1/Content/SharpFont.dll.config > SharpFont.dll.config
-	rm -rf SharpFont.2.5.0.1
+	nuget install SharpFont -Version 3.0.1
```
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps.sh]: You can then add ` -ExcludeVersion` version here as we learned in https://github.com/OpenRA/OpenRA/pull/7907.

Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps.sh]: Nevermind.

**Comment by Mailaender** [comment]
Please remove the `zlib1.dll` dependencies, too then:

* https://github.com/OpenRA/OpenRA/blob/fc0ad15a5bfe111947435eb3b1a1fb2e1ff41425/thirdparty/fetch-thirdparty-deps.sh#L44

* https://github.com/OpenRA/OpenRA/pull/7696/files#diff-39647d23f86a8d5cbb96e0b2f10ab325R31

File: thirdparty/fetch-thirdparty-deps.sh [hunk 4]
```diff
@@ -39,10 +39,13 @@ fi
...
+	cp ./SharpFont.3.0.1/lib/net20/SharpFont* .
+	cp ./SharpFont.3.0.1/config/SharpFont.dll.config .
+	rm -rf SharpFont.3.0.1
+	# No exact version of the dependencies are required on NuGet.
+	# Directory name can change because of version.
```
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps.sh]: Yes, but we don't want to do version changes by accident like in the middle of a release stabilization so please chose a matching version for us and install it before SharpFont grabs their own.

**Comment by Robmaister** [comment]
Huh, guess I wasn't looking too closely and the zlib dependency popped back up somewhere in the rebase.

Removed it and also specified SharpFont.Depdnencies.2.5.5.1.

**Comment by Mailaender** [comment]
Thanks!

File: thirdparty/fetch-thirdparty-deps-windows.sh [hunk 5]
```diff
@@ -16,10 +16,9 @@ fi
 
 if [ ! -f windows/freetype6.dll ]; then
 	echo "Fetching FreeType2 from nuget"
-	nuget install freetype2.redist -Version 2.4.11.3
-	cp ./freetype2.redist.2.4.11.3/bin/win32/zlib1.dll ./windows/
-	cp ./freetype2.redist.2.4.11.3/bin/win32/freetype6.dll ./windows/
-	rm -rf freetype2.redist.2.4.11.3
+	nuget install SharpFont.Dependencies -Version 2.5.5.1
```
Comment (Mailaender) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: Can you add `-ExcludeVersion` here and below? See https://github.com/OpenRA/OpenRA/pull/7907. It is also the reason this needs a rebase now. Sorry.

Comment (Robmaister) [inline thirdparty/fetch-thirdparty-deps-windows.sh]: Yup, actually makes a lot of sense.

**Comment by Robmaister** [comment]
Looks like the AppVeyor build failed because NSIS failed to download, can't find any way to re-run it like I can with my own projects...

**Comment by Mailaender** [comment]
Rebuilding.

**Comment by penev92** [comment]
Well it works on Windows.

**Comment by pchote** [comment]
:+1:, but there's going to be a lot of pain once this is merged as people will need to manually delete and refetch dependencies.

**Comment by Mailaender** [comment]
> but there's going to be a lot of pain once this is merged as people will need to manually delete and refetch dependencies.

I addressed this in https://github.com/OpenRA/OpenRA/pull/7953 by extending `make clean`.

**Pull request merged**